### PR TITLE
Convert user manual completely to asciidoctor

### PR DESCRIPTION
**Pull request by eriwen**
This change replaces all embedded Docbook syntax from the user manual. Asciidoctor files are now converted directly to HTML and PDF. 

It also not only greatly simplifies the user manual generation pipeline, but it makes it faster. In addition, the userguide_single page and the PDF have been given some much needed love. They now render much more cleanly.

I did the following for all user manual pages:

 - [x] Adjusted links to other manual pages, DSL docs, and Javadocs where necessary. Fixed a couple dozen broken links at least.
 - [x] Convert samples XML from docbook to asciidoctor syntax
 - [x] Converted wide tables to a more flexible list-based display

Important contributor differences from the docs as they stand before this PR:

 - Users must use a `<file>.adoc#` prefix for all cross-reference links. This is _required_ to achieve working links in all 3 outputs: multi-page user manual, single-page user manual, and PDF.
 - The manual PDF is no longer themed, and no longer has the list of examples links in the TOC.
 - File names were changed from camelCased to snake_case to have a crystal clear 1-1 mapping between source to output.
 - Section reference verification is now handled by Asciidoctor verbose mode. 
 - The single-page user manual has the same navigation as the multi-page manual for now.
 - [Common Asciidoctor extensions](https://github.com/gradle/dotorg-docs/tree/master/docs-asciidoctor-extensions) are now used by Gradle Guides and the User Manual, however, the the web assets (CSS, JS, etc) are not yet merged. These extensions are responsible for injecting the navigation and styles. The header/footer/navigation HTML can now be found under `subprojects/docs/src/main/resources`.

Here are the differences in the -all distribution from Gradle 4.9, as generated through `diff --brief -r gradle-4.9/ gradle-4.10-20180716070000+0000 | grep "Only in" | grep -E "(docs|samples)/"`. Comments are mine.

```
# All of these are expected. The CSS files are new. 
Only in gradle-4.10-20180716070000+0000/docs/dsl: manual.css
Only in gradle-4.10-20180716070000+0000/docs/dsl: prettify.css
Only in gradle-4.10-20180716070000+0000/docs/userguide: getting-started.html
Only in gradle-4.10-20180716070000+0000/docs/userguide: index.html
Only in gradle-4.10-20180716070000+0000/docs/userguide: manual.css
Only in gradle-4.10-20180716070000+0000/docs/userguide: prettify.css

# Images are now in images/ instead of img/. The former is the Asciidoctor default.
Only in gradle-4.10-20180716070000+0000/docs/userguide: images
Only in gradle-4.9/docs/userguide: img
Only in gradle-4.9/docs/userguide: introduction.html

# Parts are no longer generated. This is okay, as these aren't linked to hardly at all anywhere. We'll redirect them to the docs home page.
Only in gradle-4.9/docs/userguide: pt01.html
Only in gradle-4.9/docs/userguide: pt02.html
Only in gradle-4.9/docs/userguide: pt03.html
Only in gradle-4.9/docs/userguide: pt04.html
Only in gradle-4.9/docs/userguide: pt05.html
Only in gradle-4.9/docs/userguide: pt06.html
Only in gradle-4.9/docs/userguide: pt07.html
Only in gradle-4.9/docs/userguide: pt08.html
Only in gradle-4.9/docs/userguide: pt09.html
Only in gradle-4.9/docs/userguide: pt10.html

# No more samples list, which was only used for the PDF and single user manual. We shall have users navigate samples differently soon
Only in gradle-4.9/docs/userguide: sample_list.html

# I straight up took this page away because it's an embarrassment and is better covered by the Java Web Getting Started Guide
Only in gradle-4.9/docs/userguide: web_project_tutorial.html

# Affected by a separate PR
Only in gradle-4.10-20180716070000+0000/samples/userguide/tutorial/antLoadfile: antLoadfileResources
Only in gradle-4.9/samples/userguide/tutorial: antLoadfileResources
Only in gradle-4.10-20180716070000+0000/samples/userguide/tutorial/antLoadfileWithMethod: antLoadfileResources
```

#### Links

Links are validated in 2 ways:

 1. I manually went through and checked that every link actually links to the intended content, fixing more than several. Given an invalid reference, Docbook would happily link to the current page, which doesn't _fail_ for link checkers but it's not correct. I used the Broken Link Checker Chrome extension to check all links, fixing those that came up.
 2. Asciidoctor verbose mode will render warnings if any of the internal links break.

### Context 
This change allows us to have multi-language samples for Groovy and Kotlin, and is required for Gradle 4.10.

There will surely be follow-on polishing, such as implementing structured data, better link checking, and upgrading Asciidoctor to be more incremental and cacheable again.

### Review Guidance

While I did my very best to ensure that everything renders or links to or includes everything we expect, I need some more eyes looking in different areas to verify.

 - [ ] The online userguide. That it renders properly for all chapters.
 - [ ] The offline userguide. Users should have a reasonably good experience from the -all distribution
 - [ ] The bin/ distribution contains what we expect
 - [ ] The PDF renders as good or better than the current one. 
 - [ ] The automation doesn't leave anything out-of-place.

### CODE REVIEW

File: buildSrc/subprojects/build/src/main/groovy/org/gradle/build/docs/ExtractSamplesTask.groovy [hunk 1]
```diff
@@ -59,6 +59,55 @@ class ExtractSamplesTask extends DefaultTask {
         }
     }
 
+    /**
+     * Given an XML document, Generate HOCON .conf files in the proper location
+     */
+    private def generateSampleConf(Document doc) {
```
Comment (eriwen) [inline buildSrc/subprojects/build/src/main/groovy/org/gradle/build/docs/ExtractSamplesTask.groovy]: TODO (@eriwen): This whole class is now unnecessary. Remove it.

File: subprojects/docs/docs.gradle [hunk 2]
```diff
@@ -67,6 +69,9 @@ configurations {
 }
 
 dependencies {
+    asciidoctor "org.gradle:docs-asciidoctor-extensions:0.0.1"
```
Comment (eriwen) [inline subprojects/docs/docs.gradle]: TODO @eriwen finalize use of these extensions and publish publicly

File: subprojects/docs/docs.gradle [hunk 3]
```diff
@@ -447,12 +450,76 @@ task distDocs(type: Docbook2Xhtml) {
...
+}
+
+
 task userguide {
     dependsOn userguideHtml, userguideSingleHtml, userguidePdf
     description = 'Generates the userguide'
     group = 'documentation'
 }
 
+asciidoctor {
```
Comment (eriwen) [inline subprojects/docs/docs.gradle]: TODO @eriwen: Clean up this whole `docs.gradle` bit and convert it to Kotlin DSL

File: subprojects/docs/docs.gradle [hunk 4]
```diff
@@ -447,12 +450,76 @@ task distDocs(type: Docbook2Xhtml) {
...
+        'ant-manual-url'      : 'https://ant.apache.org/manual',
+        'guides-url'          : 'https://guides.gradle.org',
+        'gradle-version'      : version,
+        'groovy-dsl-path'     : "../dsl",
+        'javadoc-path'        : "../javadoc",
+        'samples-path'        : "../../samples"
+}
+
+task userManualPdf {
+    // TODO: user manual single to PDF
```
Comment (eriwen) [inline subprojects/docs/docs.gradle]: TODO @eriwen PDF conversion

**Comment by lptr** [comment]
Why are all the `*.adoc` files new? We've already converted them a while ago.

**Comment by eriwen** [comment]
@lptr I want to build both old and new side-by-side until the conversion is done. At that time we'll rename them back (ish) to have a better diff.

File: buildSrc/subprojects/build/build.gradle.kts [hunk 5]
```diff
@@ -1,5 +1,9 @@
 dependencies {
     compile("commons-lang:commons-lang:2.6")
     compile("com.google.guava:guava-jdk5:14.0.1")
-    compile("org.asciidoctor:asciidoctor-gradle-plugin:1.5.6")
+    compile("org.asciidoctor:asciidoctorj-pdf:1.5.0-alpha.11") {
+        exclude(module="asciidoctorj")
```
Comment (bigdaz) [inline buildSrc/subprojects/build/build.gradle.kts]: I'm sure we can do better than this with the new dependency resolution features. What's the reason for this `exclude`?

File: subprojects/docs/src/docs/userguide/build_cache.adoc [hunk 6]
```diff
@@ -100,85 +99,85 @@ BUILD SUCCESSFUL
...
 If the previous build outputs have been already stored in the cache by someone else, e.g. your continuous integration server or other developers, you can avoid executing most tasks locally.
 
-The following inputs contribute to the build cache key for a task in the same way that they do for <<sec:how_does_it_work,up-to-date checks>>:
+The following inputs contribute to the build cache key for a task in the same way that they do for <<more_about_tasks.adoc#sec:how_does_it_work,up-to-date checks>>:
 
 * The task type and its classpath
 * The names of the output properties
-* The names and values of properties annotated as described in <<sec:task_input_output_annotations>>
-* The names and values of properties added by the DSL via api:org.gradle.api.tasks.TaskInputs[]
+* The names and values of properties annotated as described in <<more_about_tasks.adoc#sec:task_input_output_annotations,the section called "Custom task types">>
```
Comment (bigdaz) [inline subprojects/docs/src/docs/userguide/build_cache.adoc]: "the section called" is not required for this link.

**Comment by lptr** [comment]
Here's a change that I almost missed among all that noise.

image

- [ ] Using `api:org.gradle.api.AntBuilder[]` looks more clear to me than the new `link:{groovy-dsl-path}/org.gradle.api.AntBuilder.html[AntBuilder]` notation. Why change this? I'd like us to keep the `api:...[]` notation unless there's good reason to change it.

- [ ] Also, `api:...[]` used to take the user to the Javadoc, not to the Groovy DSL doc. Why this change? Shouldn't we prefer Javadoc _more_ now that we also have Kotlin to support?

**Comment by wolfs** [comment]
@lptr `api:` converts to the docbook `<apilink>` tag which points to the DSL reference or the Javadoc, depending which is available. I guess @eriwen went for a simple `link:` to not make the PR even bigger.

**Comment by lptr** [comment]
@pledbrook your reasoning about preferring the DSL reference makes sense to me. I'm fine with that direction. However, I still think we should keep the `api:...[]` style syntax as it expresses the intentions better (I want to refer to _some_ documentation about this API), and it's also shorter and simpler, so fewer things to get wrong.

**Comment by eriwen** [comment]
I will first try a new PR that just renames .adoc files and hope to get that merged first so that the diff is easier to follow. I'm not sure I can easily split out the samples inlining. What I can possibly do is have a few PRs with several chapters each to make each one sane, and merge them all at the same time. 

I know there are a bunch of improvements that can be made here, and my hope is to get an initial set of changes merged first and then do polishing in yet another PR. 

Regarding `api:`, I would like to consider that a separate change. The Kotlin DSL factors in here, and I know that it's not going to become a part of the existing DSL reference, which makes things a bit more complex if we want to link to it instead of the Groovy DSL. I would rather have a couple separate inlinemacros that make writing links to Javadoc and DSL refs have nicer syntax, but again I think that's polish we can add afterward.

Thanks for the feedback so far. Please keep it coming so I can get this merged soon as reasonably possible.

File: buildSrc/subprojects/build/build.gradle.kts (hunk 5 above)
Comment (eriwen) [reply buildSrc/subprojects/build/build.gradle.kts]: Sorry, copied this from one of the [asciidoctor-pdf issues](https://github.com/asciidoctor/asciidoctor-pdf/issues) (I forget which). Anyway, I don't think this is necessary anymore. The intent was to enforce a particular version of asciidoctorj run by all tools.

**Comment by pledbrook** [comment]
@eriwen Why is it necessary to include the filename in cross references? This will make it difficult to move sections between chapters as well as make the cross references verbose.

**Comment by eriwen** [comment]
Thanks for the feedback @pledbrook. I'll fix the variables and adjust the styles roughly as suggested. 

I think the single-page user manual should probably have a longer TOC in the preamble just like the PDF, because it just won't fit on the right.

Regarding the multi-page cross references, the only solution right now for "chunked" output seems to be exactly what we were doing before: converting to Docbook XML and then converting that to HTML. 

Here's the context I found when researching this:
 - https://github.com/asciidoctor/asciidoctor-gradle-plugin/issues/52
 - http://discuss.asciidoctor.org/Recommendations-for-multi-page-and-single-page-documentation-td118.html

I'm afraid we must have a pipeline that omits Docbook in order to achieve the web experience we need to have Kotlin DSL snippets. I spent days trying to get it to do what I needed with no success.

I'm using [this method of cross-linking](https://asciidoctor.org/docs/asciidoc-writers-guide/#cross-references) here to allow links to work in all 3 outputs.

**Comment by eriwen** [comment]
Chapter references were broken by 05397b145bf02f1dbb97caa29ffc0af1756bbcdf. Reverting that fixes them again.

**Comment by eriwen** [comment]
@pledbrook You're right about the self-doc references. I'll change those to just `<<#foo, ...>>`

**Comment by eriwen** [comment]
Closing in favor of "smaller" PRs #6006 and #6021

**Pull request closed without merging**